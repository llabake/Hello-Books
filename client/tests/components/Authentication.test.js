import React from 'react';
import { Redirect } from 'react-router-dom';

import requireAuthentication, { mapStateToProps } from '../../components/Authentication';
import userData from '../mocks/userData'

let ProtectedRouteComponent, WrapperComponent;
let wrapper;
test('Authentication component should render correctly', () => {
  localStorage.setItem('token', 'tokentoken')
  ProtectedRouteComponent = () => (
    <div>
      I am a protected component
    </div>
  )
  WrapperComponent = requireAuthentication(ProtectedRouteComponent);
  const props = {
    authenticated: false,
    user: {},
    errors: {},
    authenticateUser: jest.fn(() => Promise.resolve(1)),
    dispatch: jest.fn(),
    location: {
      state: {
        from: {}
      }
    },
  }
  wrapper = shallow(
    <WrapperComponent.WrappedComponent {...props} />
  )
  expect(wrapper.exists()).toBe(true);
})

test('Authentication component calls authenticate user when component recieves props', () => {
  localStorage.setItem('token', 'tokentoken')
  ProtectedRouteComponent = () => (
    <div>
      I am a protected component
      </div>
  )
  WrapperComponent = requireAuthentication(ProtectedRouteComponent);
  const props = {
    authenticated: true,
    user: userData.user,
    errors: {},
    authenticateUser: jest.fn(() => Promise.resolve(1)),
    dispatch: jest.fn(),
    location: {
      pathname: '/profile',
    }
  }
  const nextProps = {
    authenticated: true,
    user: userData.user,
    errors: {},
    authenticateUser: jest.fn(() => Promise.reject(1)),
    dispatch: jest.fn(),
    location: {
      pathname: '/profile',
    },
    isAuthenticated: true
  }
  wrapper = shallow(
    <WrapperComponent.WrappedComponent {...props} />
  )
  expect(wrapper.exists()).toBe(true);

  let spy;
  afterEach(() => {
    spy.mockClear();
  });

  spy = jest.spyOn(WrapperComponent.WrappedComponent.prototype, 'componentWillReceiveProps')
  wrapper.setProps(nextProps);
  expect(WrapperComponent.WrappedComponent.prototype.componentWillReceiveProps).toHaveBeenCalledTimes(1);
  expect(spy).toHaveBeenCalled();
  expect(wrapper.find("Redirect")).toHaveLength(1);
  expect(wrapper).toMatchSnapshot();
})

test('Authentication component calls map state to props', () => {
  localStorage.setItem('token', 'tokentoken')
  ProtectedRouteComponent = () => (
    <div>
      I am a protected component
      </div>
  )
  WrapperComponent = requireAuthentication(ProtectedRouteComponent);
  const props = {
    authenticated: false,
    user: {},
    errors: {},
    authenticateUser: jest.fn(() => Promise.resolve(1)),
    dispatch: jest.fn(),
    location: {
      pathname: '/signin',
    }
  }
  const store = {
    error: {},
    user: userData.admin,
    authenticated: true
  }
  wrapper = shallow(
    <WrapperComponent.WrappedComponent {...props} />
  )
  expect(mapStateToProps({ userReducer: {} })).toHaveProperty('user');
  expect(mapStateToProps({ userReducer: {} })).toHaveProperty('authenticated');
  expect(mapStateToProps({ userReducer: {} })).toHaveProperty('error');
});





