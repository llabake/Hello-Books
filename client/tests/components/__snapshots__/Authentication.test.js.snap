// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication component calls authenticate user when component recieves props 1`] = `
<div>
  <Redirect
    push={false}
    to={
      Object {
        "pathname": "/signin",
        "state": Object {
          "from": Object {
            "pathname": "/profile",
          },
        },
      }
    }
  />
</div>
`;
